### PR TITLE
Calculate emission_limit from scalars and filter additional scalars

### LINKED ISSUE #163: Add calculation of emission limit
Now the emission limit is provided with the scalars.

We have just discussed that we might want to use a range of emission reductions. 

todos:
- [x] adapt input data --> add reduction factor and an amount by which the emission limit is reduced due to sectors that are not included e.g. aircraft
- [x] add function that calculates the emission limit

proposal for names:

- `emissions_1990`
- `emission_reduction_factor` e.g. 0.8 or 0.95
- `emission_reduction_amount` unit e.g. kgCO2_eq

@jnnr do you agree?

### PR DESCRIPTION
**Pull request by SabineHaas**
Fix #163

This PR needs to merged into  `feature/constraint_equate_vars` as changes of that PR are needed.

To test this PR you need to download the current scalar data. 

Note: To use the emission constraint with  `2050-el_eff`, set `scenario_key`of one of the emission reductions to `2050-base`.

### CODE REVIEW

File: scripts/build_datapackage.py [hunk 1]
```diff
@@ -242,21 +242,52 @@ def parametrize_sequences(edp, ts, filters):
...
-    return df
+    # filter by 'scenario_key'
+    for id, filt in filters.items():
+        filtered = df.copy()
+        for key, value in filt.items():
+            filtered = filter_df(filtered, key, value)
+
+    # calculate emission limit and add to data frame
+    filtered_ = filtered.copy().set_index("var_name")
+    emission_limit = (
```
Comment (jnnr) [inline scripts/build_datapackage.py:271]: This is the actual calculation. Could be a function `calculate_emission_limit` for better readability.

Comment (jnnr) [inline scripts/build_datapackage.py:271]: This is the actual calculation. Could be a function `calculate_emission_limit` for better readability.

File: scripts/build_datapackage.py [hunk 2]
```diff
@@ -242,21 +242,52 @@ def parametrize_sequences(edp, ts, filters):
...
+        {
+            "var_name": "emission_limit",
+            "var_value": emission_limit,
+            "carrier": "emission",
+            "var_unit": "kg_CO2_eq",
+        },
+        index=[0],
+    )
+
+    add_scalars = pd.concat([filtered, emission_limit_df], sort=False)
```
Comment (jnnr) [inline scripts/build_datapackage.py:286]: Why does `filtered` have to be included in the addtional scalars?

Comment (jnnr) [inline scripts/build_datapackage.py:286]: Why does `filtered` have to be included in the addtional scalars?

**Review by jnnr — COMMENTED** [review]
Hi @SabineHaas, thanks for your work! I have some questions and remarks that came up when trying to understand what is going on. Hope it helps!

File: scripts/build_datapackage.py (hunk 2 above)
Comment (SabineHaas) [reply scripts/build_datapackage.py:286]: `filtered_df` contains the (by scenario_key) filtered additional scalars and  the emission limit is included here (as it was calculated from the input data)

File: scripts/build_datapackage.py [hunk 3]
```diff
@@ -242,21 +242,52 @@ def parametrize_sequences(edp, ts, filters):
...
+    emissions = scalars.loc[scalars.carrier == EMISSION]
+
     # get `output_parameters` of backpressure components as they are not take into
     # consideration in oemof.tabular so far. They are added to the components' output flow towards
     # the heat bus in script `optimize.py`.
     bpchp_out = scalars.loc[
         (scalars.tech == "bpchp") & (scalars.var_name == "output_parameters")
     ]
+
     # save all dataframes in one file
```
[the remark this answers is not in the document]
Comment (SabineHaas) [reply scripts/build_datapackage.py]: adapted comment in [3a4275d](https://github.com/rl-institut/oemof-B3/pull/195/commits/3a4275de4591fe4458c739adb23f0c23b1122f46)

File: scripts/build_datapackage.py (hunk 2 above)
Comment (SabineHaas) [reply scripts/build_datapackage.py]: I have added a comment in [3a4275d](https://github.com/rl-institut/oemof-B3/pull/195/commits/3a4275de4591fe4458c739adb23f0c23b1122f46)

File: scripts/build_datapackage.py [hunk 4]
```diff
@@ -242,21 +242,52 @@ def parametrize_sequences(edp, ts, filters):
...
-def load_additional_scalars(scalars):
-    """Loads additional scalars like the emission limit."""
+def load_additional_scalars(scalars, filters):
+    """Loads additional scalars like the emission limit and filters by 'scenario_key'.
+    Emission limit is calculated by
+    emission_limit = (emissions_1990 - emissions_not_modeled) * (1 - emission_reduction_factor)"""
     # get electricity/gas relations and emission limit
     el_gas_rel = scalars.loc[scalars.var_name == EL_GAS_RELATION]
-    emissions = scalars.loc[scalars.var_name == EMISSION_LIMIT]
+    emissions = scalars.loc[scalars.carrier == EMISSION]
```
[the remark this answers is not in the document]
Comment (SabineHaas) [reply scripts/build_datapackage.py:251]: I have adapted the comment to make it clearer in [3a4275d](https://github.com/rl-institut/oemof-B3/pull/195/commits/3a4275de4591fe4458c739adb23f0c23b1122f46)

File: scripts/build_datapackage.py [hunk 5]
```diff
@@ -242,21 +242,52 @@ def parametrize_sequences(edp, ts, filters):
...
 
-    return df
+    # filter by 'scenario_key'
+    for id, filt in filters.items():
+        filtered = df.copy()
+        for key, value in filt.items():
+            filtered = filter_df(filtered, key, value)
+
+    # calculate emission limit and add to data frame
+    filtered_ = filtered.copy().set_index("var_name")
```
[the remark this answers is not in the document]
Comment (SabineHaas) [reply scripts/build_datapackage.py]: done in [f1262fb](https://github.com/rl-institut/oemof-B3/pull/195/commits/f1262fb6b94f4cb2cfdaada7ecc19466bc8c40e5)

File: scripts/build_datapackage.py [hunk 6]
```diff
@@ -242,21 +242,52 @@ def parametrize_sequences(edp, ts, filters):
...
+
     # save all dataframes in one file
     df = pd.concat([el_gas_rel, emissions, bpchp_out])
 
-    return df
+    # filter by 'scenario_key'
+    for id, filt in filters.items():
+        filtered = df.copy()
+        for key, value in filt.items():
+            filtered = filter_df(filtered, key, value)
```
[the remark this answers is not in the document]
Comment (SabineHaas) [reply scripts/build_datapackage.py]: Thanks for seeing this. Concatenating the data frames was missing and is added in [f1262fb](https://github.com/rl-institut/oemof-B3/pull/195/commits/f1262fb6b94f4cb2cfdaada7ecc19466bc8c40e5).

I keep both of the loops like in `parametrize_scalars` as I want the same filter options, i.e. the data could be filtered by additional columns.

File: scripts/build_datapackage.py (hunk 1 above)
Comment (SabineHaas) [reply scripts/build_datapackage.py]: done in [3e4cac9](https://github.com/rl-institut/oemof-B3/pull/195/commits/3e4cac95f6595530d4544eba4f2a2e85422d0c33)

**Review by henhuy — APPROVED** [review]
Code-wise this looks okay to me.

**Comment by jnnr** [comment]
Hi @henhuy, after merging dev into this branch, I got a failing test in `test_config`. Could you have a look?

**Review by jnnr — APPROVED** [review]
Thanks for your work, @SabineHaas! I refactored the filtering such that the subsequent filtering uses the same function as in the parametrization of the EnergyDatapackage. Tested it with all scenarios and saved the results.

**Pull request merged**